lib/internal/modules/esm/loader.js

```diff
@@ -202,33 +211,74 @@ class ESMLoader {
       const { ModuleWrap, callbackMap } = internalBinding('module_wrap');
       const module = new ModuleWrap(url, undefined, source, 0, 0);
       callbackMap.set(module, {
-        importModuleDynamically: (specifier, { url }) => {
-          return this.import(specifier, url);
+        importModuleDynamically: (specifier, { url }, importAssertions) => {
+          return this.import(specifier, url, importAssertions);
         }
       });
 
       return module;
     };
     const job = new ModuleJob(this, url, evalInstance, false, false);
     this.moduleMap.set(url, job);
+    finalFormatCache.set(job, 'module');
     const { module } = await job.run();
 
     return {
       namespace: module.getNamespace(),
     };
   }
 
-  async getModuleJob(specifier, parentURL) {
+  async getModuleJob(specifier, parentURL, importAssertions) {
```

Comment:
@bmeck’s biggest concern with the previous version, which I share, was that it broke the one-to-one relationship between cache entries and module jobs (https://github.com/nodejs/node/pull/39921/files#r714998813). One of you can please correct me, but isn’t that issue still present as long as we have `importAssertions` as part of this function signature? As in, there could be different module jobs returned based on the import assertions. Maybe not currently as the function is written, if all `importAssertions` is used for is error checking, but the signature seems to strongly imply that a different module job might be returned based on different input which includes the import assertions.

Could we remove `importAssertions` from this signature and do the assertion checking in the calling function, `import`? Or is there a better way to keep clear that there is a one-to-one relationship between resolved specifiers and module jobs?

Reply:
The spec mandates that while "no assertions" and "any arbitrary set of assertions" may return the same module record, and any of them can error, there can never be two module records produced by those. In other words, the implication you're saying the signature makes here is one the spec prohibits.

Additionally, the spec permits two different sets of assertions (including "nothing") may produce the same module record.

(i don't think any of this contradicts your comment, but it seemed worth clarifying)

Reply:
@ljharb we are not saying it is prohibited by the spec, but that it causes some issues if done in reality, just like how ESM is allowed to run completely synchronously in the spec and was designed to do so, but doesn't in reality once attempting implementation. The cache behavior is gnarly to figure out how to handle around user loaders (in particular allowing for web behavior via user loader and both assert and non-assert behavior in the same API space) but probably could be designed around given time unlike the async -> sync issue.

Reply:
no implementation issues exist to my knowledge outside of the cache behaviors on collision at this time.

Reply:
That's great to hear. I'm hoping we can design all of the APIs now, with the assumption that assertionless JSON imports will be possible later alongside assertioned ones, so that we're not boxing ourselves into a corner wrt that decision later.

In other words, I understand that the current decision is "expose only assertioned JSON imports, for now", but until the decision is made to NEVER allow assertionless JSON imports, it seems better to write all the code assuming it will be allowed in the future.